### ISSUE TITLE
Bug: [consistent-indexed-object-style] Fails on circular references involving multiple types

### ISSUE THREAD

**Issue by Peeja** [typescript-eslint/typescript-eslint#7148]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=PTAElByAoAXBPAHApqAKgJ0cgvKA3qAbQGtFYAuUAZ2jQEsA7AcwF0L6BXAWwCNE1QAPqgzIAvgG5IkEKEAy5DATJ0mAIygc+YqQrU6TVqA48+g4ZgBMoCQqRnEljYRLkqNBizZde-IcsRqJIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra1mAE0QAPRMNocARgCtEZOn0JJ0UAO4BDaE0jgwAXxCGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
// ✅
type Tree = { [key: string]: number | Tree };

// ❌
type Tree1 = { [key: string]: number | Tree2 };
type Tree2 = { [key: string]: number | Tree1 };
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/consistent-indexed-object-style": "warn"
  },
};
```

### tsconfig

```jsonc
{}
```

### Expected Result

The rule should not report and attempt to fix `Tree1` and `Tree2`.

### Actual Result

The rule reports `Tree1` and `Tree2`, and tries to `--fix` it to the following, which causes TS errors:

```ts
// 2456: Type alias 'Tree1' circularly references itself. 5:6 - 5:11
type Tree1 = Record<string, number | Tree2>;
// 2456: Type alias 'Tree2' circularly references itself. 6:6 - 6:11
type Tree2 = Record<string, number | Tree1>;
```

### Additional Info

This is the same as #2687 (fixed in #4347), except that there's now more than one type alias involved. The solution there was to check whether the type was defined recursively (referring to itself directly), but that doesn't tell us whether the type is defined *circularly* through multiple types.

That's clearly a harder problem to solve, and there may be no sensible resolution.

This arose while trying to type JSON values as:

```ts
type JsonPrimitive = string | number | boolean | null;
type JsonArray = JsonValue[];
type JsonObject = { [key: string]: JsonValue};
type JsonValue = JsonPrimitive | JsonArray | JsonObject;
```

The rule wants to change `JsonObject` to `Record<string, JsonValue>`, which causes a circular reference.

**Comment by Peeja** [comment]
(Whoops, forgot to add the title.) 🙈

**Comment by bradzacher** [comment]
This is a bit of a "really hard problem to solve" and a bit of "limitation with TS" (#41164)

In order to solve this we need to inspect every single index signature and mapped type to extract the names that it references, then inspect those referenced names in the same way to determine if there are any circular references.

It's doable, for sure, but it is prohibitively expensive to do this amount of work given the rarity of circularly referenced types whose loop is indirect [1]. I believe it would require a massive refactor of the rule to handle.

As much as I would love to fix this - it's something we're going to have to leave as an edge case that we can't support.

I don't know if it's even worth documenting this edge case given how rarer it is!

---

[1] rarity in that you're the first one to report the a big due to indirect circularity - it's just not something people do much.

**Comment by Peeja** [comment]
Hmm. Given that defining a JSON type hits it, I'm a bit surprised it's that rare. But rare or not, I definitely see why it would be a lot of work to address.

Can we lean on TypeScript itself? TS is already able to report the circularity. Can we ask it if a change would cause circularity, and skip it if it would?

**Comment by bradzacher** [comment]
> Given that defining a JSON type hits it, I'm a bit surprised it's that rare

Most people don't define generic json types! It's a rare usecase to do procesing of unstructured json data - most people either assume correct structure (eg `parse as Type`) or they use a lib like zod to do the safe parsing for them.

---

> Can we lean on TypeScript itself? TS is already able to report the circularity.

No - this requires access to type information.

The only way to truly handle this as well is with type information - because the cycle could be hidden behind imports across modules.

But given the rarity of the necessity to have cyclic types that are also an indexed type - making such a breaking change to the rule just isn't worth it due to the fact that it completely cuts off so many users from using the rule.
